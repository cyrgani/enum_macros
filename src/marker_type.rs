use alloc::vec;
use proc_macro2::{Span, TokenStream};
use quote::{format_ident, quote};
use syn::parse::{Parse, ParseStream};
use syn::punctuated::Punctuated;
use syn::{Fields, Generics, ItemEnum, Meta, Result, Token, Variant};

pub struct Arguments {
    attrs: Punctuated<Meta, Token![,]>,
}

impl Parse for Arguments {
    fn parse(input: ParseStream) -> Result<Self> {
        Ok(Self {
            attrs: input.parse_terminated(Meta::parse, Token![,])?,
        })
    }
}

pub fn marker_type(args: Arguments, item: ItemEnum) -> TokenStream {
    let marker = ItemEnum {
        variants: item
            .variants
            .iter()
            .map(|variant| Variant {
                fields: Fields::Unit,
                ..variant.clone()
            })
            .collect(),
        ident: format_ident!("{}Marker", item.ident, span = Span::call_site()),
        generics: Generics::default(),
        attrs: vec![],
        ..item.clone()
    };

    let attrs = args
        .attrs
        .into_iter()
        .map(|meta| {
            quote! {
                #[#meta]
            }
        })
        .collect::<TokenStream>();

    quote! {
        #item

        #[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
        #attrs
        #marker
    }
}
