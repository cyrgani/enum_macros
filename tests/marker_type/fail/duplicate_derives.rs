// The error output of this test will improve once https://github.com/rust-lang/rust/issues/131083 is resolved.
use enum_macros::marker_type;

#[marker_type(derive(Clone, Copy, Debug, Hash, PartialEq, Eq))]
enum Complex {
    A(bool),
    B { val: u8 },
}

fn main() {}
