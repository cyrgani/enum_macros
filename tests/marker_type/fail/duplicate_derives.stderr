error[E0119]: conflicting implementations of trait `Clone` for type `ComplexMarker`
 --> tests/marker_type/fail/duplicate_derives.rs:4:22
  |
4 | #[marker_type(derive(Clone, Copy, Debug, Hash, PartialEq, Eq))]
  | ---------------------^^^^^-------------------------------------
  | |                    |
  | |                    conflicting implementation for `ComplexMarker`
  | first implementation here
  |
  = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `Copy` for type `ComplexMarker`
 --> tests/marker_type/fail/duplicate_derives.rs:4:29
  |
4 | #[marker_type(derive(Clone, Copy, Debug, Hash, PartialEq, Eq))]
  | ----------------------------^^^^-------------------------------
  | |                           |
  | |                           conflicting implementation for `ComplexMarker`
  | first implementation here
  |
  = note: this error originates in the derive macro `Copy` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `Debug` for type `ComplexMarker`
 --> tests/marker_type/fail/duplicate_derives.rs:4:35
  |
4 | #[marker_type(derive(Clone, Copy, Debug, Hash, PartialEq, Eq))]
  | ----------------------------------^^^^^------------------------
  | |                                 |
  | |                                 conflicting implementation for `ComplexMarker`
  | first implementation here
  |
  = note: this error originates in the derive macro `Debug` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `Hash` for type `ComplexMarker`
 --> tests/marker_type/fail/duplicate_derives.rs:4:42
  |
4 | #[marker_type(derive(Clone, Copy, Debug, Hash, PartialEq, Eq))]
  | -----------------------------------------^^^^------------------
  | |                                        |
  | |                                        conflicting implementation for `ComplexMarker`
  | first implementation here
  |
  = note: this error originates in the derive macro `Hash` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `StructuralPartialEq` for type `ComplexMarker`
 --> tests/marker_type/fail/duplicate_derives.rs:4:48
  |
4 | #[marker_type(derive(Clone, Copy, Debug, Hash, PartialEq, Eq))]
  | -----------------------------------------------^^^^^^^^^-------
  | |                                              |
  | |                                              conflicting implementation for `ComplexMarker`
  | first implementation here
  |
  = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `PartialEq` for type `ComplexMarker`
 --> tests/marker_type/fail/duplicate_derives.rs:4:48
  |
4 | #[marker_type(derive(Clone, Copy, Debug, Hash, PartialEq, Eq))]
  | -----------------------------------------------^^^^^^^^^-------
  | |                                              |
  | |                                              conflicting implementation for `ComplexMarker`
  | first implementation here
  |
  = note: this error originates in the derive macro `PartialEq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `Eq` for type `ComplexMarker`
 --> tests/marker_type/fail/duplicate_derives.rs:4:59
  |
4 | #[marker_type(derive(Clone, Copy, Debug, Hash, PartialEq, Eq))]
  | ----------------------------------------------------------^^---
  | |                                                         |
  | |                                                         conflicting implementation for `ComplexMarker`
  | first implementation here
  |
  = note: this error originates in the derive macro `Eq` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `*self` which is behind a shared reference
 --> tests/marker_type/fail/duplicate_derives.rs:4:22
  |
4 | #[marker_type(derive(Clone, Copy, Debug, Hash, PartialEq, Eq))]
  |                      ^^^^^ move occurs because `*self` has type `ComplexMarker`, which does not implement the `Copy` trait
  |
note: if `ComplexMarker` implemented `Clone`, you could clone the value
 --> tests/marker_type/fail/duplicate_derives.rs:5:1
  |
4 |   #[marker_type(derive(Clone, Copy, Debug, Hash, PartialEq, Eq))]
  |                        ----- you could clone this value
5 | / enum Complex {
6 | |     A(bool),
7 | |     B { val: u8 },
8 | | }
  | |_^ consider implementing `Clone` for this type
  = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0507]: cannot move out of `*self` which is behind a shared reference
 --> tests/marker_type/fail/duplicate_derives.rs:4:1
  |
4 | #[marker_type(derive(Clone, Copy, Debug, Hash, PartialEq, Eq))]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because `*self` has type `ComplexMarker`, which does not implement the `Copy` trait
  |
note: if `ComplexMarker` implemented `Clone`, you could clone the value
 --> tests/marker_type/fail/duplicate_derives.rs:5:1
  |
4 |   #[marker_type(derive(Clone, Copy, Debug, Hash, PartialEq, Eq))]
  |   --------------------------------------------------------------- you could clone this value
5 | / enum Complex {
6 | |     A(bool),
7 | |     B { val: u8 },
8 | | }
  | |_^ consider implementing `Clone` for this type
  = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)
